awx/main/models/jobs.py

```diff
@@ -836,6 +838,27 @@ def get_notification_templates(self):
     def get_notification_friendly_name(self):
         return "Job"
 
+    def get_hosts_for_fact_cache(self):
+        """
+        Builds the querset to use for writing or finalizing the fact cache
+        these need to be the 'real' hosts associated with the job.
+        For constructed inventories, that means the original (input inventory) hosts
+        when slicing, that means only returning hosts in that slice
+        """
+        Host = JobHostSummary._meta.get_field('host').related_model
+        if not self.inventory_id:
+            return Host.objects.none()
+
+        if self.inventory.kind == 'constructed':
+            id_field = Host._meta.get_field('id')
+            host_qs = Host.objects.filter(id__in=Host.objects.exclude(instance_id='').values_list(Cast('instance_id', output_field=id_field)))
+        else:
+            host_qs = self.inventory.hosts
+
+        host_qs = host_qs.only(*HOST_FACTS_FIELDS)
+        host_qs = self.inventory.get_sliced_hosts(host_qs, self.job_slice_number, self.job_slice_count)
+        return host_qs
```

Comment:
@relrod this is ultimately derived from the method you had introduced. Took me a while but it's making more sense to me. For a given job, there's a list of potential hosts that it may touch, so this is that. There's an exception case for constructed inventory as well as for job slicing.